sql/000002.sql

```diff
@@ -0,0 +1,183 @@
+      values(
+        identifier,
+        coalesce(payload, '{}'),
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        coalesce(run_at, now()),
+        coalesce(max_attempts, 25)
+      )
+      returning *
+      into v_job;
+  else
+    insert into graphile_worker.jobs (task_identifier, payload, queue_name, run_at, max_attempts, key)
+      values(
+        identifier,
+        coalesce(payload, '{}'),
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        coalesce(run_at, now()),
+        coalesce(max_attempts, 25),
+        job_key
+      )
+      on conflict (key) do update set
+        -- update job details if provided, otherwise maintain existing value
+        -- note we can't use excluded here, or the value will always be updated,
+        -- due to the coalesce statements in values ( ... ) above
+        task_identifier=coalesce(add_job.identifier, jobs.task_identifier),
+        payload=coalesce(add_job.payload, jobs.payload),
+        queue_name=coalesce(add_job.queue_name, jobs.queue_name),
+        max_attempts=coalesce(add_job.max_attempts, jobs.max_attempts),
+
+        -- update run_at if argument is provided. If not, and there has been an
+        -- error, assume run_at reflects the error retry backoff, so reset it.
+        -- If no errors, maintain existing value by default
+        run_at=coalesce(
+          add_job.run_at,
+          (case when jobs.attempts > 0
+            then now()
+            else jobs.run_at
+          end)
+        ),
+
+        -- always reset error/retry state
+        attempts=0,
+        last_error=null
+      where jobs.locked_at is null
+      returning *
+      into v_job;
+
+    -- if the returned id is null, assume job is already locked for processing
+    -- and couldn't be updated
+    if v_job.id is null then
+      -- remove existing key to allow a new one to be inserted, and prevent any
+      -- subsequent retries by bumping attempts to the max allowed
+      update graphile_worker.jobs set
+        key = null,
+        attempts = jobs.max_attempts
+      where key = job_key;
+
+      -- insert the new job. Assume no conflicts due to the update above
+      insert into graphile_worker.jobs(task_identifier, payload, queue_name, run_at, max_attempts, key)
+        values(
+          identifier,
+          coalesce(payload, '{}'),
+          coalesce(queue_name, public.gen_random_uuid()::text),
+          coalesce(run_at, now()),
+          coalesce(max_attempts, 25),
+          job_key
+        )
+        returning *
+        into v_job;
+    end if;
+  end if;
+  return v_job;
+end;
+$$ language plpgsql;
+
+
+--- implement new remove_job function
+
+create function graphile_worker.remove_job(
+  job_key text
+) returns graphile_worker.jobs as $$
+  delete from graphile_worker.jobs
+    where key = job_key
+    and locked_at is null
+  returning *;
+$$ language sql;
```

Comment:
In this case if `job_key` is null then the result is null so we can skip the function call (`STRICT`) and it's volatile which I prefer to be explicit about (since it's the default, I prefer that we explicitly state we decided it SHOULD be volatile, rather than that it's just volatile because we weren't paying close enough attention).

```suggestion
$$ language sql strict volatile;
```

Could you make similar edits to the other functions if necessary please? Very few will be `strict`, mind.

Reply:
Have done - I've made `fail_job` strict too, seeing as none of its arguments have defaults. The other functions have optional arguments which default to null.